### ISSUE TITLE
When importing assets into sub-folder of volume, the assets end up in the wrong folder

### ISSUE THREAD

**Issue by rob-baker-ar** [craftcms/feed-me#581]
### Description

Given a JSON feed with substantial number of rows (more than 300) like this:

```JSON
[
    ...,
    { 
        "url": "https://www.example.com/?no=path&but=plenty&of=parameters", 
        "filename": "filename.jpg",
        "folder": "folder/structure/4-levels/deep"
    },
    ...
]
```

Each folder in the folder path is more likely to be different than the same as the last, creating a large sub-folder tree into which the file(s) should be saved. Unfortunately, the folders are created usually incorrectly.

For example, given this JSON: `{ "filename": "BE0647-walk-03.jpg", "folder": "print/BE/0/BE0647" }` the file "BE0647-walk-03.jpg" (or in fact not that filename on disk due to #580) ending up in the folder `original/AN/7/BE0647/`.

This issue _feels_ like some sort of cache or hashing issue where it is not checking the filesystem for every parent / child folder relationship.

### Additional info

- Plugin version: 4.1.1
- Craft version: 3.2.5.1

**Comment by brandonkelly** [comment]
Same response as https://github.com/craftcms/feed-me/issues/580#issuecomment-520079567 .

Feed Me only lets you specify the URL to fetch the images from, but not which folder they should be saved to. If they are getting saved into subfolders, that should be a result of your Assets field’s Upload Location setting.

**Comment by rob-baker-ar** [comment]
Similar comment to #580 ...

![feedme filename example](https://user-images.githubusercontent.com/39988088/62853211-8961d900-bce3-11e9-86bf-abd20502c26a.png)

The screen shot shows filename and folder field mapped from the feed.

Am I mis-understanding what "filename" and "folder" mappings are for in an Asset based import?

**Comment by rob-baker-ar** [comment]
The issue here may have been to do the the folder structure, as in the feed.

From the Snippet of JSON in the initial report at the top, data in reality looks like:

```
...
"folder": "FR/0/FR0292",
...
"folder": "FR/3/FR3291",
...
```

The issue seems to occur when FeedMe creates folders with single digit folder names - in that case, it doesnot create the folder, does not generate an error and uploads the file to a different previously created folder (seemingly random, there probably is a pattern but I cannot see it at the moment).

**Comment by floatingbits** [comment]
Created a pull request
https://github.com/craftcms/feed-me/pull/1053

**Comment by i-just** [comment]
This has been working as expected for a while now, so I’m going to close this issue. 

I tested against v4.4.0 (the latest one at the moment is 4.8.0) and v5.0.4 (the latest one at the moment is 5.3.0). In both cases, 2 assets from external URLs were to be imported with a custom filename and with a custom folder (`2021/08/foo` for the first asset and `2021/08/bar` for the second one) - both imported as expected.

**Comment by amphibian** [comment]
Actually, I don't think this is fixed, as I'm encountering this issue when importing a WordPress media library (using an Assets import).

WordPress stores media in date-based folders using the format `YYYY/MM`. This results in folders with matching names, but nested at different levels (e.g., `2021/08`, `2022/08`, `2023/08`, etc).

Feed Me is not distinguishing between the folder named `08` that lives in the `2023` folder, and the folder named `08` that lives in the `2021` folder. So I'm ending up with a total of 12 imported folders only (one for each month), and images that should end up in `2023/08` or `2022/08` instead go `2021/08` (since it was created first and returns as a matching final folder name).

This explains why, in the original issue, `print/BE/0/BE0647` ended up in `original/AN/7/BE0647` - I bet `original/AN/7/BE0647` was pre-existing, and the final folder - `BE0647` - is a match. But in your test, @i-just, `foo` and `bar` have distinct names.

**Comment by rob-c-baker** [comment]
In the end, I worked around this (a long time ago!) by modifying the imported folder structure so there were no folder names that started with a `0` and had no "digit only" names. In our case that also meant we achieved unique parent folder structure too.

So in the original example:

`print/BE/0/BE0647` was transformed to `print/BE/BE0/BE0647`

That's maybe not quite what @amphibian is describing and to be honest - with a large dataset - as I can attest to, is difficult to achieve.

We did it by: Not actually changing the folder structure, but instead wrote a PHP script to deliver the images that would be accessible over HTTP on the server that held the images which delivered the files, but appearing in the URL as the transformed folder structure. Feedme took the path part of the URL and used that instead of the on-disk / real folder structure.

Luckily, this was a one-off import and the apparatus to make it work has long since been taken down.
